pitop/processing/algorithms/ball_detect.py

```diff
@@ -0,0 +1,269 @@
+from collections import deque
+import numpy as np
+from pitop.core import ImageFunctions
+from pitop.processing.core.vision_functions import (
+    center_reposition,
+    get_object_target_lock_control_angle,
+)
+from typing import Union
+from imutils import resize, grab_contours
+from imutils.video import FPS
+from os import getenv
+from pitop.core.data_structures import DotDict
+import atexit
+from pitop.processing.core.vision_functions import (
+    import_opencv,
+    tuple_for_color_by_name
+)
+
+valid_colors = ["red", "green", "blue"]
+
+color_ranges = {
+    "red": [
+        {
+            "lower": (150, 100, 100),
+            "upper": (179, 255, 255)
+        },
+        {
+            "lower": (0, 100, 100),
+            "upper": (5, 255, 255)
+        }
+    ],
+    "green": [{
+        "lower": (60, 100, 100),
+        "upper": (90, 255, 255)
+    }],
+    "blue": [{
+        "lower": (100, 100, 100),
+        "upper": (130, 255, 255)
+    }]
+}
+
+MIN_BALL_RADIUS = 5
+BALL_CLOSE_RADIUS = 50
+DETECTION_POINTS_BUFFER_LENGTH = 64
+
+
+class Ball:
+    def __init__(self, color):
+        self.color = color
+
+        # Red needs to be a higher limit
+        # since some skin types are redish in color
+        match_limits = {
+            "red": 0.1,
+            "green": 0.05,
+            "blue": 0.05
+        }
+        self.match_limit = match_limits[color]
+
+        self.center_points = deque(maxlen=DETECTION_POINTS_BUFFER_LENGTH)
+        self.radius = 0
+        self.angle_from_center = None
+
+    def clear(self):
+        self.center_points.clear()
+        self.radius = 0
+        self.angle_from_center = None
+
+    @property
+    def angle(self):
+        return self.angle_from_center
+
+    @property
+    def center(self):
+        if len(self.center_points) > 0:
+            return self.center_points[0]
+        return None
+
+    def is_valid(self):
+        return self.center is not None
+
+
+class BallDetector:
+    def __init__(self,
+                 process_image_width: int = 320,
+                 format: str = "OpenCV",
+                 print_fps: bool = False):
+        """
+        :param int process_image_width: image width to scale to for image processing
+        :param str format: output image format
+        """
+        self.cv2 = import_opencv()
```

Comment:
Why does this need to be here? If this algorithm is being imported then we know it is needed.
Simply putting `cv2 = import_opencv()` at the top would do it, right?

Reply:
I was told to put it as a class member to prevent cv2 errors if someone tries to import another file that triggers one of the __init__ files to import this one as a side effect